provide basic bind backend

anthonysomerset opened the pull request:
Hi This PR provides basic Bind backend support

1) will install the correct backend package like other modules
2) the DB parameters are now optional so they are not required for this backend (logic might need to be tightened up in other backends to check and "require" them.)
3) still need to set `backend_install` and `backend_create_tables` to false (yet to make these parameters optional)

I am working on a seperate resource type to "create" zones at the very least for bind will open a separate PR once i have completed this more.

ju5t commented:
I've changed the merge request to target the develop branch. 

Let me know when you're ready for a review.

anthonysomerset commented:
I'm just about ready to review - i had fixed the initial build and formatting errors and introduced some basic tests for the new backend too - theres a test error on puppet 5 now but i'm not 100% sure how the test can be improved accordingly because the parameter is now optional to support bind backend

ju5t commented:
The test expects the error message to contain:

```
parameter 'db_username' expects a String\[1, default\] value, got String
```

But the error message is:

```
Error while evaluating a Resource Statement, Class[Powerdns]: parameter 'db_username' expects a value of type Undef or String[1, default], got String
```

The test should be:

```
expect { subject.call } .to raise_error(/parameter 'db_username' expects a value of type Undef or String\[1, default\], got String/)
```

Instead of

```
expect { subject.call } .to raise_error(/parameter 'db_username' expects a String\[1, default\] value, got String/)
```

ju5t commented on manifests/backends/bind.pp:
```diff
@@ -0,0 +1,48 @@
+    group  => 'root',
+  }
+
+  file_line { 'powerdns-bind-baseconfig':
```
This resource is not tested yet. Coverage dropped to 98.53%.

anthonysomerset replied:
this item should now be fixed

anthonysomerset commented:
Test coverage updated - now covering 100% of resources

Slightly annoyingly - that DB test fix *only* works on Puppet 5, this test now produces errors on puppet versions < 5

i'm not sure we want to bump minimum puppet version to 5 over a test - but not sure how to make this test work between differing puppet versions accordingly

ju5t commented:
I didn't expect that.

You could test per Puppet version too but it seems a bit much. It's easier to improve the regex:

```
expect { subject.call } .to raise_error(/parameter 'db_username' expects a (value of type Undef or )?String\[1, default\]( value)?, got String/)
```

I've tested this on Puppet 4.7 and 5.

anthonysomerset commented:
I've just made these final changes - tweaking grammer and fixing the test cases

have also changed names to named.conf to better align with manuals and /etc/pdns/named

ju5t commented:
@anthonysomerset I have nothing else to add. If you want I can merge this into our develop branch.

anthonysomerset commented:
yes please - i've got nothing more on this at this stage, thanks

ju5t commented:
Thanks for the new feature. I'll release 1.3.1 later this week.
